src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -787,6 +789,49 @@ void primitive_inst::do_runtime_skip_reorder() {
     }
 }
 
+void primitive_inst::do_runtime_skip_gather() {
+    // Target pattter:
+    // [kv_cache_input / beam_idx_input] => Gather => Concat
+    auto is_optimizable_gather = [&] {
+        if (!get_node().is_type<gather>())
+            return false;
+        if (!get_node().get_dependency(0).is_type<input_layout>() &&
+            !get_node().get_dependency(0).is_type<read_value>())
+            return false;
+        if (get_user_insts().size() != 1 || !get_user_insts().front()->get_node().is_type<concatenation>())
```

Comment:
@vladimir-paramuzov not sure if we can make it generic. AFAIK, there are no any requirements for dictionary data repeatlessness for gather operation. But for beam search we may consider that all of indexes are unique, right? But it can't be a common assumption without additional indexes check

Reply:
> But for beam search we may consider that all of indexes are unique, right?

No, indices may repeat for beam search. We can optimize out gather in case when 
1. indices is 1d tensor
2. `indices.shape[0] == dictionary.shape[axis]` - gather doesn't change shape
3. `indices.data == [0, 1, ..., indices.shape[0]]` - and doesn't modify elements order along axis

In that case input tensor is not modified. And from my pov that part doesn't depend on any surrounding operation

Reply:
Got it, thanks for clarification! I missed that indexes check already implemented and performs unconditionally... Certainly, we can make it generic in such a case. Sorry for the confusion

Reply:
@vladimir-paramuzov  The reason why I limited the target pattern to stateful one is that the "checking indices value" might cause another overhead for which is not likely to be the target case. How do you think?

Reply:
That's fair, though maybe we can replace primitive type check with some shape limit, i.e. if indices.shape[axis] > some_limit, then disable optimization. In that case it's more clear that optimization is generic, but there are some perf aspects to consider.

Reply:
Okay, I will come up with better condition.

Reply:
Fixed